File: src/main/java/Parser.java
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * Author: stk
 * Date: 18/4/1
 */
public class Parser {
    public static void main(String[] args) {
        String projectPath = "C:\\Users\\stk\\Documents\\Docs\\Papers\\EASE2018Log\\Data\\elasticsearch";
        String outputPath = "C:\\Users\\stk\\Downloads\\result.txt";
        Parser parser = new Parser();
        parser.generate(parser.parseProject(projectPath), outputPath);
    }

    public List<String> parseProject(String path) {
        try {
            return Files.walk(Paths.get(path))
                    .filter(p -> p.toString().endsWith(".java"))
                    .map(p -> FileParser.parseFile(p.toString()))
                    .filter(Objects::nonNull)
                    .flatMap(Collection::stream)
                    .collect(Collectors.toList());
        } catch (IOException e) {
            e.printStackTrace();
        }
        throw new IllegalArgumentException();
    }

    public void generate(List<String> logs, String path) {
        if (new File(path).delete()) System.out.println("Delete existing result file.");
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(path))) {
            logs.forEach(s -> {
                try {
                    writer.write(s);
                    writer.write("\n");
                } catch (IOException e) {
                    e.printStackTrace();
                }
            });
            writer.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/scala/Parser.scala
import java.io.{File, PrintWriter}
import java.util.concurrent.ConcurrentHashMap

import scala.collection.mutable
import scala.io.Source

/**
  * Author: stk
  * Date: 18/4/6
  */
object Parser {
  val root = "F:\\GitHub Data\\"
  val name = "100-hadoop"

  def main(args: Array[String]): Unit = {
    val projectPath = root + name
    val outputPath = root + name + ".txt"
    val files = getBlameFiles(new File(projectPath)).toVector
    //    generate(parseProject(files), outputPath)
    sortProject(files)
  }

  def getBlameFiles(file: File): Array[File] = {
    val files = file.listFiles
    files.filter(_.isFile).filter(_.toString.endsWith(".blame")) ++ files.filter(_.isDirectory).flatMap(getBlameFiles)
  }

  def parseProject(files: Vector[File]): Vector[String] = files.par.map(FileParser.parseFile).reduce(_ ++ _)

  def sortProject(files: Vector[File]): Unit = {
    val container = new ConcurrentHashMap[String, Int]()
    val seq = files.par.map(sort).reduce((m1, m2) => m1 ++ m2.map { case (k, v) => k -> (v + m1.getOrElse(k, 0)) }).toSeq.sortBy(_._2).reverse
    val writer = new PrintWriter(new File(root + name + "-loc.txt"))
    seq.foreach(f => writer.write("%-10d\t%s\n".format(f._2, f._1)))
    writer.flush()
    writer.close()
  }

  def sort(file: File): mutable.HashMap[String, Int] = {
    println("Sort file: " + file)
    val regex = "^[\\^\\w]+\\s*(.*)\\s*\\((.*)\\s+\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2} .\\d{4}\\s+(\\d+)\\) (.*)$".r
    val source = Source.fromFile(file)
    val lines = source.getLines
    val map = new mutable.HashMap[String, Int]()
    lines.foreach(f => {
      regex.findFirstMatchIn(f).foreach(p => {
        val name = p.group(2).trim
        if (map.contains(name))
          map += (name -> (map(name) + 1))
        else
          map += (name -> 1)
      })
    })
    source.close()
    map
  }

  def generate(logs: Vector[String], path: String): Unit = {
    if (new File(path).delete) println("Delete existing result file.")
    val writer = new PrintWriter(new File(path))
    logs.foreach(s => writer.write(s + "\n"))
    writer.flush()
    writer.close()
  }
}
